### PR TITLE
Revamp to use startgg Python package and improve GHA workflows

### LINKED ISSUE #16: Current test GHA workflow is very inefficient
It's probably best to just run the regular tests in the standard python:3.11-slim docker image rather going through the whole build process for every push. Even with caching it's excessive.

Additionally, the runtime Docker image can probably avoid installing gcc if we just build all the wheels in the requirements-stage and move them over or something

### PR DESCRIPTION
**Pull request by charlesepwilson**
- Uses the new custom startgg package to allow for API responses to have a known data structure. This also moves to async functions for the requests, which should be an improvement.
- Fixes dynamic versioning
- Adds ruff , black and docformatter to improve code standardisation. Switches from mypy to pyright for discord.py compatibility reasons
- Improved logging
- Added deployment workflow (yet to be seen whether this actually works though...)
- 
Resolves #16 #15 #12 #9 #13
#14 might be resolved. Improved logging has been added in case not.

### CODE REVIEW

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+name: Deploy
```
Comment (tolxe) [inline .github/workflows/deploy.yml:1]: one small question - why not just have a singular `definitions.yaml` file which contains both the contents of `build-and-test.yml` and `deploy.yml`. Is it just for the sake of clarity or is there a bigger reason I'm missing 🤔

File: luke_bot/discord_bot.py [hunk 2]
```diff
@@ -1,28 +1,33 @@
...
-    """Checks if 2 blocks of update text contain the same information"""
-    lines1 = update1.strip().splitlines()
-    lines2 = update2.strip().splitlines()
+    """Checks if 2 blocks of update text contain the same information."""
+
+    def clean(update: str) -> list[str]:
+        return update.strip().splitlines()
+
+    lines1 = clean(update1)
+    lines2 = clean(update2)
```
Comment (tolxe) [inline luke_bot/discord_bot.py:24]: mainly a reminder for myself but should probably rename this to something more informative than just lines1 and lines2 😅

**Review by tolxe — APPROVED** [review]
looks good 👀 just added a qq for clarity

File: .github/workflows/deploy.yml (hunk 1 above)
Comment (charlesepwilson) [reply .github/workflows/deploy.yml:1]: The triggers and environments associated with these workflows are different, so I think it's best to keep them separate. Deploy only happens on a release, while build and test happens on every push.

File: luke_bot/discord_bot.py (hunk 2 above)
Comment (charlesepwilson) [reply luke_bot/discord_bot.py]: I've changed this now anyway. I've only renamed it from lines1/2 to old_lines and new_lines so if you want to re-rename things in the future with your own ideas go ahead.

**Pull request merged**